### ISSUE TITLE
2.10: `validated_data` must be provided if `call_default_factory` is True (affets: SQLModel)

### ISSUE THREAD

**Issue by zzstoatzz** [pydantic/pydantic#10912]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

this issue opened per https://github.com/pydantic/pydantic/issues/10905#issuecomment-2489657973

### Description

appears to be a new requirement that `validated_data` is passed to `Field.get_default` if `call_default_factory` is True, which was not the case before

this was introduced [here](https://github.com/pydantic/pydantic/pull/10678/files#diff-bfaaac2ae0a71d9686a272982d29d0bed82bfaacf8b5b30e2b88ed2543e74a3bR581) (see related [comment](https://github.com/pydantic/pydantic/pull/10678/files#r1815062438))

### edit

[I think I understand why this new requirement exists now
](https://github.com/pydantic/pydantic/issues/10912#issuecomment-2489782906)

---

### trace
after upgrading to 2.10 we saw the following error start popping up

```python
    def get_default(self, *, call_default_factory: bool = False, validated_data: dict[str, Any] | None = None) -> Any:
    ...
        if self.default_factory is None:
            return _utils.smart_deepcopy(self.default)
        elif call_default_factory:
            if validated_data is None:
>               raise ValueError("'validated_data' must be provided if 'call_default_factory' is True.")
E               ValueError: 'validated_data' must be provided if 'call_default_factory' is True.
```

### Example Code (MRE)

```Python
from typing import ClassVar, List, Set

from pydantic import BaseModel, Field


class PrefectBaseModel(BaseModel):
    _reset_fields: ClassVar[Set[str]] = set()

    def reset_fields(self):
        return self.model_copy(
            update={
                field: self.model_fields[field].get_default(call_default_factory=True)
                for field in self._reset_fields
            }
        )


class TaskRun(PrefectBaseModel):
    _reset_fields: ClassVar[Set[str]] = {"tags"}

    name: str
    tags: List[str] = Field(default_factory=list)


def main():
    task = TaskRun(name="test-task")
    task.tags.append("new-tag")
    task.reset_fields()


if __name__ == "__main__":
    main()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.0
        pydantic-core version: 2.27.0
          pydantic-core build: profile=release pgo=false
                 install path: /Users/nate/github.com/prefecthq/prefect/.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.7 (main, Oct  7 2024, 23:45:10) [Clang 18.1.8 ]
                     platform: macOS-15.0.1-arm64-arm-64bit
             related packages: pydantic-settings-2.6.1 mypy-1.13.0 pydantic-extra-types-2.10.0 fastapi-0.115.5 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by zzstoatzz** [comment]
this reproduces

```python
from typing import ClassVar, List, Set

from pydantic import BaseModel, Field


def default_tags() -> List[str]:
    return ["default"]


class PrefectBaseModel(BaseModel):
    _reset_fields: ClassVar[Set[str]] = set()

    def reset_fields(self):
        return self.model_copy(
            update={
                field: self.model_fields[field].get_default(call_default_factory=True)
                for field in self._reset_fields
            }
        )


class TaskRun(PrefectBaseModel):
    name: str
    tags: List[str] = Field(default_factory=default_tags)
    _reset_fields = {"tags"}


def main():
    task = TaskRun(name="test-task")
    task.tags.append("new-tag")
    task.reset_fields()


if __name__ == "__main__":
    main()
```

**Comment by zzstoatzz** [comment]
ok looking at my example, if I pass `validated_data=self.model_dump()` then _I think_ it does what I want
```python
from typing import ClassVar, List, Set

from pydantic import BaseModel, Field


class PrefectBaseModel(BaseModel):
    _reset_fields: ClassVar[Set[str]] = set()

    def reset_fields(self):
        return self.model_copy(
            update={
                field: self.model_fields[field].get_default(
                    call_default_factory=True, validated_data=self.model_dump()
                )
                for field in self._reset_fields
            }
        )


class TaskRun(PrefectBaseModel):
    _reset_fields: ClassVar[Set[str]] = {"tags"}

    name: str
    tags: List[str] = Field(default_factory=list)


def main():
    task = TaskRun(name="test-task")
    task.tags.append("new-tag")
    task.reset_fields()

    print(task)


if __name__ == "__main__":
    main()
```

```python
» python sandbox/repros/pydantic210.py
name='test-task' tags=['new-tag']
```

**Comment by zzstoatzz** [comment]
> EDIT:
ok I think I get it. https://github.com/pydantic/pydantic/pull/10678, that's what we need to pass if we're going to tell it to call_default_factory

after realizing this, I'm realizing this breaking change may have been unavoidable, but I will keep this issue open until told otherwise

**Comment by sydney-runkle** [comment]
Maybe fixed by https://github.com/pydantic/pydantic/pull/10909

**Comment by zzstoatzz** [comment]
@sydney-runkle yes I think it does! using the pre-workaround version above

```python
» uv run --with prefect==3.1.3 --with pydantic==2.10.0  sandbox/repros/pydantic210.py
Traceback (most recent call last):
  File "/Users/nate/github.com/prefecthq/prefect/sandbox/repros/pydantic210.py", line 34, in <module>
    main()
  File "/Users/nate/github.com/prefecthq/prefect/sandbox/repros/pydantic210.py", line 28, in main
    task.reset_fields()
  File "/Users/nate/github.com/prefecthq/prefect/sandbox/repros/pydantic210.py", line 12, in reset_fields
    field: self.model_fields[field].get_default(call_default_factory=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nate/github.com/prefecthq/prefect/.venv/lib/python3.12/site-packages/pydantic/fields.py", line 596, in get_default
    raise ValueError("'validated_data' must be provided if 'call_default_factory' is True.")
ValueError: 'validated_data' must be provided if 'call_default_factory' is True.
» uv run --with prefect==3.1.3 --with pydantic@git+https://github.com/pydantic/pydantic.git@validated-data-fix sandbox/repros/pydantic210.py
 Updated https://github.com/pydantic/pydantic.git (69e509ef)
name='test-task' tags=['new-tag']
```

**Comment by joshuamckenty** [comment]
Just wanted to chime in that this broke creating new instances of SqlModels in the last few hours; will try and read through to understand how to fix it.

**Comment by bubblegumsoldier** [comment]
Any update on this? Broke our whole system. Let me know if we can help.

**Comment by sydney-runkle** [comment]
> Any update on this? Broke our whole system. Let me know if we can help. Anyone knows the last working pydantic version?

We should be releasing a patch today, hoping to get a fix in for this one! Should have more info soon.

**Comment by sydney-runkle** [comment]
Hey folks! We've just released [v2.10.1](https://github.com/pydantic/pydantic/releases/tag/v2.10.1) with a fix for this issue!

Let us know if you're still experiencing any difficulties. Thanks!

**Comment by zzstoatzz** [comment]
> Let us know if you're still experiencing any difficulties. Thanks!

fwiw - things [lgtm here](https://github.com/PrefectHQ/prefect/pull/16087/files), thank you 🙂
